enhancements/cloud-integration/infrastructure-external-platform-type.md

```diff
@@ -0,0 +1,513 @@
+
+- As a cloud provider whose platform is not integrated into OpenShift,
+  I want to have the Cloud Controller Manager for my infrastructure running in OpenShift from the initial install.
+  Having a platform type that allows for the addition of operators or components which perform platform-specific functionality would help
+  me to create better integrations between OpenShift and my infrastructure platform.
+
+### Goals
+
+- Reduce the necessity to make changes in "openshift/api", "openshift/library-go" and dependant infrastructure-related components during basic cloud-provider integration with OpenShift
+- Make a cloud provider integration process more accessible and simple to external contributors as well as for RH engineers
+- Provide a top-level description of the operation mode for affected components at the first steps
+- Introduce a somewhat "neutral" platform type, which would serve as a signal about an underlying cloud-provider presence
+
+### Non-Goals
+
+- describe concrete delivery mechanisms for cloud-provider specific components
+- cover new infrastructure provider enablement from the RHCOS side
+- describe specific changes for each affected component, aside no-op reaction to the new "External" platform type
+
+## Proposal
+
+Our main goal is to simplify a new cloud providers integration process with the OpenShift/OKD. To achieve this
+we are proposing to add new "External" [PlatformType](https://github.com/openshift/api/blob/51f399230d604fa013c2bb341040c4ad126e7309/config/v1/types_infrastructure.go#L128)
+along with respective Spec and Status structures into the "openshift/api".
+
+Such a generic platform type will serve as a signal for built-in Openshift operators about an underlying cloud-provider presence.
+Related PlatformSpec and PlatformStatus type structures will serve as a source of generic configuration information for
+the Openshift-specific operators.
+
+Having that special platform type will allow infrastructure partners to clearly designate when their OpenShift
+deployments contain components that replaces and/or supplement the core Red Hat components.
+
+### Affected components
+
+This section enumerates OpenShift's components and briefly elaborates the future plans around this proposal.
+During [initial implementation](#implementation-phases) we must ensure that all OpenShift's components treat "External" platform in the same way as
+"None" to ensure we do not disrupt the current openshift operation.
+
+In the future, we will need to change the behavior of OpenShift components to be able to receive supplemental provider-specific components from an infrastructure provider
+or, if a component manages something else (i.e. kubelet, kcm), adjust its behaviour (set `--cloud-provider=external` arg to kubelet for example).
+
+Specific component changes will be described in detail within separate enhancement documents on a per-component basis.
+
+#### Machine Config Operator
```

Comment:
another aspect of MCO which is platform dependent is the networking part. In case of on prem platforms, and vips, networking manifests are being injected which ease the installation process.
In case of External platform, what should be the desired behavior? I think we do have to find a way to allow this functionality - so maybe the solution is to decouple it from platform type and provide another knob

Reply:
> In case of External platform, what should be the desired behavior?

i would expect this to follow the pattern you mention in the first part, namely injecting extra manifests which would be processed during the install. were you thinking of something specific we should support for the external platform?

Reply:
I guess it's more about apiserver internal IPs and things like that, which we have in status for BareMetal, Openstack, Vsphere and other on-premish platforms, for example: https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L482

We do have URLs for apiserver on the level above in the InfrastructureStatus: https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L82

But yeah, i agree, this is an important aspect which i kinda missed here...

Reply:
I did think about that, couple of things which came to my head:

* DNS presence is a requirement for the openshift installation and since we have apiserver url in the [InfrastructureStatus](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L82). I think this should be enough as the starting point, since DNS server and load-balancer like things might be deployed using manifests injection during installation.

* >I think we do have to find a way to allow this functionality - so maybe the solution is to decouple it from platform type and provide another knob
I agree, this should be decoupled from the platform. For now it is very hard to me to predict how many "cloud-providers" would want such knobs (how many of them have load-balancer like solutions, internal DNS and so on).
For now i don't feel that we should add this into the PlatformStatus for the "External" platform. Think better add this as part of future work and in case if strict necessity will shown up.

*  For the "External" one this definitely should be an opting-in thingy.

Reply:
Ref: https://github.com/openshift/enhancements/blob/ce4d303db807622687159eb9d3248285a003fabb/enhancements/network/baremetal-networking.md

Reply:
+1, i agree with @lobziik 's thinking here

Reply:
It would be very useful to have in-cluster networking available for the external platform. As an example of how it's useful, the assisted-installer defaults to using the "baremetal" platform for all multi-node clusters primarily so it can use the in-cluster networking, and thus have a self-sufficient cluster regardless of the actual underlying infrastructure. How can we enable something similar when selecting the "external" platform"?

A provider will have a significantly easier time adopting OpenShift with the external platform support if they can start by launching a viable self-sufficient cluster, then focus on adding/replacing components based on their priorities and capabilities, rather than being forced to do a bunch of integration work with load balancers and DNS just to get off the ground.

I agree that long-term it's best to decouple in-cluster networking from the platform, but can we do that in time to be available when we start encouraging actual use of the external platform?

Reply:
> I agree that long-term it's best to decouple in-cluster networking from the platform, but can we do that in time to be available when we start encouraging actual use of the external platform?

good question, i'm not sure we could complete that before people are using the external platform. ideally we would be able to do it before we have people wanting to exercise those networking options.

Reply:
@mhrivnak I agree, it would be beneficial to have this knobs for the "external" platform, couple of things  concerns me about adding this right now:
1) It lives in PlatformStatus - I do not feel that it's the right API design for this thing. I would like not to spread this further.
2) This should be a feature which, IMHO, requires explicit optin-in, and this will require some of the logic changes within MCO during phase 1 (adding the "External" type, which would be equivalent to the "None")
3) We do not know yet, how many partners/interested parties would want it. I referred this [here](https://github.com/openshift/enhancements/pull/1234/files#diff-08391a193550547988305aefa40e474c593e97d3b2375245c31015f39f5812b3R151) in the Future work section.

>How can we enable something similar when selecting the "external" platform"?


>I agree that long-term it's best to decouple in-cluster networking from the platform, but can we do that in time to be available when we start encouraging actual use of the external platform?

I think for the "External" platform, this might be decoupled from the beginning, I can imagine a separate CRD with these parameters and a knob that engages this "In Cluster Network Infrastructure" for this platform type. But, IMHO, this should be designed separately with migration strategy and procedures research for the other platforms (vsphere, openstack, baremetal).
For me, it feels like a massive chunk of design work which I'm a bit concerned about adding here now.

Reply:
Agreed. Let's see if we can get that work prioritized so it's at least ready sooner than later.

Reply:
@lobziik are we ready to resolve this thread?

Reply:
@dhellmann, I think yes.